File: citadel-first/backend/app/schemas/crs_tax_residency.py
from pydantic import BaseModel, field_validator


class CrsTaxResidencyRow(BaseModel):
    jurisdiction: str
    tin_status: str  # "have_tin" or "no_tin"
    tin: str | None = None
    no_tin_reason: str | None = None  # "A", "B", or "C"
    reason_b_explanation: str | None = None

    @field_validator("tin_status")
    @classmethod
    def validate_tin_status(cls, v: str) -> str:
        if v not in ("have_tin", "no_tin"):
            raise ValueError("tin_status must be 'have_tin' or 'no_tin'")
        return v

    @field_validator("no_tin_reason")
    @classmethod
    def validate_no_tin_reason(cls, v: str | None, info) -> str | None:
        # no_tin_reason is required when tin_status is "no_tin"
        return v


class CrsTaxResidencyRequest(BaseModel):
    """Request body for PUT /signup/crs-tax-residency — replaces all rows for the user."""
    residencies: list[CrsTaxResidencyRow]


class CrsTaxResidencyResponse(BaseModel):
    id: int
    app_user_id: int
    jurisdiction: str
    tin_status: str | None
    tin: str | None
    no_tin_reason: str | None
    reason_b_explanation: str | None

    model_config = {"from_attributes": True}


class CrsTaxResidencyListResponse(BaseModel):
    residencies: list[CrsTaxResidencyResponse]

File: citadel-first/mobile/lib/models/crs_tax_residency.dart
/// Data models for CRS tax residency rows.
class CrsTaxResidencyRow {
  String jurisdiction;
  String? tinStatus; // "have_tin" or "no_tin"
  String? tin;
  String? noTinReason; // "A", "B", or "C"
  String? reasonBExplanation;
  String? otherJurisdiction; // Used when jurisdiction is "Others"

  CrsTaxResidencyRow({
    this.jurisdiction = '',
    this.tinStatus,
    this.tin,
    this.noTinReason,
    this.reasonBExplanation,
    this.otherJurisdiction,
  });

  Map<String, dynamic> toJson() => {
        'jurisdiction': jurisdiction,
        'tin_status': tinStatus,
        'tin': tin,
        'no_tin_reason': noTinReason,
        'reason_b_explanation': reasonBExplanation,
        'other_jurisdiction': otherJurisdiction,
      };

  CrsTaxResidencyRow copyWith({
    String? jurisdiction,
    String? tinStatus,
    String? tin,
    String? noTinReason,
    String? reasonBExplanation,
    String? otherJurisdiction,
  }) =>
      CrsTaxResidencyRow(
        jurisdiction: jurisdiction ?? this.jurisdiction,
        tinStatus: tinStatus ?? this.tinStatus,
        tin: tin ?? this.tin,
        noTinReason: noTinReason ?? this.noTinReason,
        reasonBExplanation: reasonBExplanation ?? this.reasonBExplanation,
        otherJurisdiction: otherJurisdiction ?? this.otherJurisdiction,
      );
}